### ISSUE TITLE
synchronize doesn't work when target is a VM on localhost on a different port

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29553]
From @ffledgling on 2016-10-26T10:25:21Z

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

synchronize
##### ANSIBLE VERSION

```
ansible 2.1.2.0
  configured module search path = Default w/o overrides

```
##### CONFIGURATION

```
[defaults]
inventory = hosts
callback_plugins = callback
```
##### OS / ENVIRONMENT

Linux -> Linux
Controller: Fedora 22
Target: Fedora 22
##### SUMMARY

synchronize refuses to copy files from the controller to the target machine, instead copying files around on the controller machine
##### STEPS TO REPRODUCE

```
[user@controller]$ echo 'Hi World' > /tmp/foo
[user@controller]$ ansible-playbook test.yaml
...
[fedora@vm]$ ls /tmp/bar
ls: cannot access /tmp/bar: No such file or directory
...
[user@controller]$ )$ cat /tmp/bar 
Hi World
```

`test.yaml`

```
- hosts: vm
  gather_facts: no

  tasks:
    - name: rsync a file
      synchronize: src=/tmp/foo dest=/tmp/bar

```

`hosts` file:

```
[vm]
localhost ansible_port=10022 ansible_user=fedora ansible_ssh_port=10022 ansible_ssh_user=fedora
```

##### EXPECTED RESULTS

Expected the file to be copied over to the target machine, which is a VM running on localhost (on a separate port) in this instance.
##### ACTUAL RESULTS

rsync/synchronize entirely ignores, in no particular order:
- The target machine (vm)
- The target user (fedora)
- The target ssh port (10022)

Instead choosing to do a local copy of the files, assuming that the source and destination are the same machine, which they are not.

```
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: ajaisingh
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1477476574.41-266884575643295 `" && echo ansible-tmp-1477476574.41-266884575643295="` echo $HOME/.an
sible/tmp/ansible-tmp-1477476574.41-266884575643295 `" ) && sleep 0'
<localhost> PUT /tmp/tmpAxswMG TO /home/ajaisingh/.ansible/tmp/ansible-tmp-1477476574.41-266884575643295/synchronize
<localhost> EXEC /bin/sh -c 'chmod u+x /home/ajaisingh/.ansible/tmp/ansible-tmp-1477476574.41-266884575643295/ /home/ajaisingh/.ansible/tmp/ansible-tmp-1477476574.41-2668845
75643295/synchronize && sleep 0'
<localhost> EXEC /bin/sh -c 'LANG=en_US.utf8 LC_ALL=en_US.utf8 LC_MESSAGES=en_US.utf8 /usr/bin/python /home/ajaisingh/.ansible/tmp/ansible-tmp-1477476574.41-266884575643295/synchroni
ze; rm -rf "/home/ajaisingh/.ansible/tmp/ansible-tmp-1477476574.41-266884575643295/" > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {"changed": true, "cmd": "/usr/bin/rsync --delay-updates -F --compress --archive --rsh 'ssh  -S none -o StrictHostKeyChecking=no -o Port=10022' --out-format='<<CHANGED
>>%i %n%L' \"/tmp/foo\" \"/tmp/bar\"", "invocation": {"module_args": {"_local_rsync_path": "rsync", "_substitute_controller": false, "archive": true, "checksum": false, "compress": true, "cop
y_links": null, "delete": false, "dest": "/tmp/bar", "dest_port": 10022, "dirs": false, "existing_only": false, "group": null, "links": null, "mode": "push", "owner": null, "partial": false, 
"perms": null, "private_key": null, "recursive": null, "rsync_opts": null, "rsync_path": null, "rsync_timeout": 0, "set_remote_user": true, "src": "/tmp/foo", "ssh_args": null, "times": null,
 "verify_host": false}}, "msg": ">f+++++++++ foo\n", "rc": 0, "stdout_lines": [">f+++++++++ foo"]}
```

Copied from original issue: ansible/ansible-modules-core#5385

**Comment by ansibot** [comment]
From @ansibot on 2016-10-26T10:25:21Z



@ansible ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @ffledgling on 2016-10-26T10:25:21Z

Further info, I've tried the latest version of ansible, and the problem persists with the `synchronize` module, but the `copy` module works correctly/as expected.

I tested this against:

```
$ ansible --version
ansible 2.3.0 (devel d59034ca68) last updated 2016/10/26 15:59:52 (GMT +550)
  lib/ansible/modules/core: (detached HEAD c51ced56cc) last updated 2016/10/26 16:23:59 (GMT +550)
  lib/ansible/modules/extras: (detached HEAD 8ffe314ea5) last updated 2016/10/26 16:24:14 (GMT +550)
  configured module search path = Default w/o overrides
```

**Comment by ansibot** [comment]
From @tima on 2016-10-26T10:25:21Z

@ffledgling I think the problem may be that you're using localhost that's not a recommended practice. Because of this it seems local is winning out over making a remote connection as you want. Have you tried forcing the connection type to smart or ssh? Have you tried giving this alias a different name like local_vm? 

needs_info

**Comment by ansibot** [comment]
From @ansibot on 2016-10-26T10:25:21Z

@ffledgling ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @ffledgling on 2016-10-26T10:25:21Z

@tima I don't exactly follow, where am I using localhost? I have a specific host called `vm` defined in my hosts file that I'm using from my playbook. Which alias would you like me to give a different name? `local_vm` instead of `vm` ?

Is the suggestion to not use VMs running on localhost in the first place?

**Comment by ansibot** [comment]
From @tima on 2016-10-26T10:25:21Z

@ffledgling In this host file you have:

```localhost ansible_port=10022 ansible_user=fedora ansible_ssh_port=10022 ansible_ssh_user=fedora```

You're using an alias called localhost that is overriding the localhost host name. Not sure that is negatively effecting things.

**Comment by ansibot** [comment]
From @ffledgling on 2016-10-26T10:25:21Z

@tima That partly effects things, but that's not the real reason for the error. From what I've read in synchronize's module code, it tries to optimize for the `localhost` case by checking the target address `.startswith('localhost')` and then strips/does not add `user@hostname:` before the destination path for `rsync`, thus trying to "save" an ssh connection. Which basically means if you're trying to ssh into a localhost for whatever reason, it just won't work. Ideally it should check for the connection type and at least in the case where `ansible_connection` is set to `ssh` or similar it should not try to optimize the `rsync` command. Ideally the optimization logic can be improved, but I think the `ansible_connection` type should always override the optimization based on it's value's semantics.

I really don't think the optimization for the local connection should lie inside the module code. It should really be based on the value of `ansible_connection`, if it's set to `local`, then the optimization makes sense, in most other cases there are semantic edge cases that just don't work.

**Comment by ansibot** [comment]
From @tima on 2016-10-26T10:25:21Z

@ffledgling The purpose of the synchronize module is to make **the most common** tasks quick and easy, not **all** the things rsync can do possible. What you're suggesting sounds outside of that scope of common and makes things less easy for most users. I don't see this as a bug.

The work around is to not use "localhost" in you inventory. Aliasing it as something else should work just fine.

wontfix

**Comment by ansibot** [comment]
From @ffledgling on 2016-10-26T10:25:21Z

@tima Is copying files from the host machine onto a VM a common task not considered a common task?

Nothing I'm suggesting tries to add additional rsync features to the module. I'm not even suggesting a change in behaviour for existing scripts, ideally a correctly implemented version of this change is opaque to existing users. How does this make it less easy for most users?

What I'm suggesting is that the synchronize module shouldn't try to optimize/change behaviour based on the name of the host. That's inconsistent and in a case such as the one above, incorrect. It should instead use the value of the `ansible_connection` for whatever it's doing, so that it's consistent with what the rest of ansible does instead of having module specific snow-flake behaviour, isn't that what `ansible_connection` is intended for? 

Sure aliasing localhost to something else and then using it works, but that's not obvious to anyone who's not read the code for the module itself. The "localhost optimization" isn't documented and the resulting behaviour is difficult to debug. It's also inconsistent with how other related ansible modules behave, for example the fetch and copy modules don't have this/similar issues and do the right thing™.

Is this really a wontfix? If you're absolutely opposed to not changing the hacky optimization, this should at the very least be documented behaviour?

**Comment by ansibot** [comment]
From @tima on 2016-10-26T10:25:21Z

> It should instead use the value of the ansible_connection for whatever it's doing, so that it's consistent with what the rest of ansible does...

I don't see you using `ansible_connection` in the issue you filed.

> What I'm suggesting is that the synchronize module shouldn't try to optimize/change behaviour based on the name of the host.

We don't have a choice really. `rsync` cannot live inside of the Ansible connection system by its very nature of providing its own remote transport system so we attempt to infer and simulate how the connection subsystem would work the the best we can. Removing that connection type optimization as you suggest would be a change in the behavior of that module.

The common usage is work is to not call a VM localhost like you are here. If you name that VM something like "default" set the inventory hostname to something like 127.0.0.1 then you have the desired behavior. Here is the unit test for that: https://github.com/ansible/ansible/blob/devel/test/units/plugins/action/fixtures/synchronize/basic_vagrant/task_vars_in.json.

Your welcome to submit a PR for review.
